### ISSUE TITLE
Implementation: `Integrator`

### ISSUE THREAD

**Issue by piljoong-jeong** [piljoong-jeong/MLX-NeRF#1]
`Integrator` is a sequence of components to **integrate** radiance fields from selected dataset. Typically, `Integrator` can hold components such as:
* Encoders
* MLPs
* Samplers
* Renderers (& Losses)

This structure is highly inspired from [`nerfstudio`'s implementation](https://github.com/nerfstudio-project/nerfstudio/blob/main/nerfstudio/models/vanilla_nerf.py#L83). However, care must be taken to properly construct automatic differentiation graph of `mlx`. Due to the lack of tensor manipulation implementation in `mlx` (e.g., `searchsorted` not implemented, etc), those unimplemented features should be done via other tensor libraries (e.g., `numpy`, `torch`, etc), with casting `mlx.array`s back and forth. `mlx` fails to construct computational graph whenever those array conversion is included in loss calculation function (which is fed into `mlx.nn.value_and_grad`), hence we have to properly separate the actual loss calculation function and the auxiliary functionalities.

In short, 
* We cast `mx.array` into `torch.tensor` to use `torch.searchsorted`, as there's no such implementation in `mlx`
* If those conversion is included in the loss calculation, `mlx` throws error
* This can be alleviated by changing typical NeRF loss calculation procedure

From:
```python
@mx.compile
def loss(batch_rays, gt):
    z_vals = uniform_sampling(batch_rays)
    view = # from `batch_rays`
    pos = # cam. aligned `z_vals`

    rgb, alpha = coarse.train(pos, view)
    rgb_coarse, weights, ... = render(...)

    # FIXME: here raises error
    z_importance_samples = sampling.sample_from_inverse_cdf(
        np.array(z_vals, copy=False), 
        np.array(weights, copy=False), 
        n_importance_samples
    )

    # ...
```

To:
```python

def integrate(batch_rays, gt):
    z_vals = uniform_sampling(batch_rays)
    view = # from `batch_rays`
    pos = # cam. aligned `z_vals`
    
    @mx.compile
    def loss_coarse(pos, view, model):
        # do the train
    loss_coarse(pos, view, coarse)
    
    # NOTE: here the error does not raised as the conversion is not included in loss function
    z_importance_samples = sampling.sample_from_inverse_cdf(
        np.array(z_vals, copy=False), 
        np.array(weights, copy=False), 
        n_importance_samples
    )
    pos_fine = # cam. aligned `z_importance_samples`
    
    @mx.compile
    def loss_fine(pos_fine, view, model):
        # do the train
    loss_fine(pos_fine, view, fine)
    
    # ...
```
